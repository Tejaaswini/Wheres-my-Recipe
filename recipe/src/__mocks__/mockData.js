const mockData = {
    signInData: {
        email: "sherlock.musk23@fmail.com",
        password: "1234qwer"
    },
    authResponse: {
        data: {
            "auth_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1NTEyODU1MjEsImlhdCI6MTUxOTc0OTUyMSwidXNlcl9pZCI6NCwiZW1haWwiOiJwYXRyaWNrLndhbHVrYWdnYUBhbmRlbGEuY29tIiwiZmlyc3RfbmFtZSI6IlBhdHJpY2siLCJsYXN0X25hbWUiOiJXYWx1a2FnZ2EiLCJwdWJsaWNfaWQiOiJiYjFkNTFlMC0xNGIzLTQyMjAtYmU0NS0zOTdhZTViZTlkOTAifQ.a624_Hg_ElxuGvqUHhhRWbhwGg4uSd9EU9Glq7oCuik",
            "message": "Successfully logged in",
            "status": "success"
        }
    },
    signUpData: {
        first_name: "musk",
        last_name: "elon",
        email: "sherlock.musk23@fmail.com",
        password: "1234qwer"
    },
    regResponse: {
        message: "Successfully registered",
        status: "success"
    },
    logoutResponse: {
        message: "User has logged out successfully.",
        status: "success"
    },
    categories: {
        "1": {id: 1, name: "Lunch", description: "Awesome lunch"},
        "2": {id: 2, name: "Dinner", description: "Awesome dinner"},
        "3": {id: 2, name: "Supper", description: "Awesome Supper"},
    },
    recipes: {
        "1": {id: 1, name: "Lunch recipe", description: "Awesome lunch", ingredients: "food, juice", directions: "Prepare well"},
        "2": {id: 2, name: "Dinner recipe", description: "Awesome dinner", ingredients: "food, juice", directions: "Prepare well"},
        "3": {id: 2, name: "Supper recipe", description: "Awesome Supper", ingredients: "food, juice", directions: "Prepare well"},
    }
}

export default mockData;