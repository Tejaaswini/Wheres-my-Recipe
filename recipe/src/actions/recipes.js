import instance from './AxiosInstance';
import { notify } from 'react-notify-toast';
import { 
    ROOT_URL, CREATE_RECIPE, FETCH_RECIPES, FETCH_RECIPE, FETCH_CATEGORY_RECIPES, DELETE_RECIPE, EDIT_RECIPE 
} from '../constants';

export const createRecipe = (values, cat_id, callback) => {
    return async (dispatch) => {
        const request = await instance.post(`${ROOT_URL}/recipe_category/${cat_id}/recipes`, values)
        .then((response) => {
            dispatch({
                type: CREATE_RECIPE,
                payload: response
            });
            callback();
            notify.show(response.data.message, 'success', 5000);
        })
        .catch((error) => {
            dispatch({
                type: "UNAUTHENTICATED",
                payload: "Invalid authentication credentials"
            });
            notify.show(error.response.data.message, 'error', 5000);
        });
    }
    
}

export const fetching = () => {
    return {
        type: "FETCHING"
    }
}

export const fetchRec = (res) => {
    return {
        type: FETCH_RECIPES,
        payload: res,
    }
}

// action creator for fetching recipes from the database
export const fetchRecipes = (value, page) => {
    return async (dispatch) => {
        
        dispatch(fetching);
        if(value){
            const limit = 4;
            let request = await instance.get(`${ROOT_URL}/search_recipes?q=${value}&limit=${limit}`)
            .then((response) => {
                dispatch(fetchRec(response));
            })
            .catch((error) => {
                dispatch({
                    type: "UNAUTHENTICATED",
                    payload: "Invalid authentication credentials"
                });
                localStorage.removeItem('current_user');
                notify.show(error.response.data.message, 'error', 5000);
            });
            if(page){
                request = await instance.get(`${ROOT_URL}/recipes?q=${value}&limit=${limit}&page=${page}`)
                .then((response) => {
                    dispatch(fetchRec(response));
                })
                .catch((error) => {
                    dispatch({
                        type: "UNAUTHENTICATED",
                        payload: "Invalid authentication credentials"
                    });
                    localStorage.removeItem('current_user');
                    notify.show(error.response.data.message, 'error', 5000);
                });
            }
        }else{
            const limit = 4;
            let request = await instance.get(`${ROOT_URL}/recipes?limit=${limit}`)
            .then((response) => {
                dispatch(fetchRec(response));
            })
            .catch((error) => {
                dispatch({
                    type: "UNAUTHENTICATED",
                    payload: "Invalid authentication credentials"
                });
                localStorage.removeItem('current_user');
                notify.show(error.response.data.message, 'error', 5000);
            });
            if(page){
                request = await instance.get(`${ROOT_URL}/recipes?limit=${limit}&page=${page}`)
                .then((response) => {
                    dispatch(fetchRec(response));
                })
                .catch((error) => {
                    dispatch({
                        type: "UNAUTHENTICATED",
                        payload: "Invalid authentication credentials"
                    });
                    localStorage.removeItem('current_user');
                    notify.show(error.response.data.message, 'error', 5000);
                });;
            }
        }
    }
}

// action creator for fetching recipes in category from the database
export const fetchCategoryRecipes = (id, page) => {
    return async (dispatch) => {
        try {
            dispatch(fetching);
            if(page){
                const limit = 4;
                const request = await instance.get(`${ROOT_URL}/recipe_category/${id}/recipes?limit=${limit}&page=${page}`);
                dispatch({
                    type: FETCH_CATEGORY_RECIPES,
                    payload: request,
                });
            }else{
                const limit = 4;
                const request = await instance.get(`${ROOT_URL}/recipe_category/${id}/recipes?limit=${limit}`);
                dispatch({
                    type: FETCH_CATEGORY_RECIPES,
                    payload: request,
                });
            }
        }catch(error) {
            dispatch({
                type: "UNAUTHENTICATED",
                payload: "Invalid authentication credentials"
            });
            localStorage.removeItem('current_user');
            notify.show(error.response.data.message, 'error', 5000);
        }
    }
}

// action creator for fetching single recipe
export const fetchRecipe = (cat_id, recipe_id) =>{
    return async (dispatch) => {
        const request = await instance.get(`${ROOT_URL}/recipe_category/${cat_id}/recipes/${recipe_id}`)
        .then((response) => {
            dispatch({
                type: FETCH_RECIPE,
                payload: response
            })
        })
        .catch((error) => {
            dispatch({
                type: "UNAUTHENTICATED",
                payload: "Invalid or expired token please login again"
            })
            localStorage.removeItem('current_user');
            notify.show(error.response.data.message, 'error', 5000);
        });
    }

}

// action creator for delete recipe
export const deleteRecipe = (cat_id, recipe_id, callback) => {
    return async (dispatch) => {
        const request = await instance.delete(`${ROOT_URL}/recipe_category/${cat_id}/recipes/${recipe_id}`)
        .then((response) => {
            dispatch({
                type: DELETE_RECIPE,
                payload: request
            });
            callback();
            notify.show(response.data.message, 'success', 5000);
        })
        .catch((error) => {
            dispatch({
                type: "UNAUTHENTICATED",
                payload: "Invalid or expired token please login again"
            })
            notify.show(error.response.data.message, 'error', 5000);
        });
    }
}

// action creator for delete recipe
export const editRecipe = (values, cat_id, recipe_id, callback) => {
    return async (dispatch) => {
        const request = await instance.put(`${ROOT_URL}/recipe_category/${cat_id}/recipes/${recipe_id}`, values)
        .then((response) => {
            dispatch({
                type: EDIT_RECIPE,
                payload: request
            });
            notify.show(response.data.message, 'success', 5000);
            callback()
        })
        .catch((error) => {
            dispatch({
                type: "UNAUTHENTICATED",
                payload: "Invalid or expired token please login again"
            })
            notify.show(error.response.data.message, 'error', 5000);
        });
    }
}