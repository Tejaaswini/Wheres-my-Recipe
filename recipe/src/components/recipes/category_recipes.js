import _ from 'lodash';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { withRouter } from 'react-router';
import Notifications from 'react-notify-toast';
import recipeListImg from '../../img/recipe_img.jpg'

import RecipeModal from '../recipes/recipe_modal';
import { fetchCategoryRecipes } from '../../actions/recipes';

class CategoryRecipes extends Component {
    componentDidMount() {
        const { id } = this.props.match.params;
        this.props.fetchCategoryRecipes(id);
    }

    reloadRecipes = () => {
        const { id } = this.props.match.params;
        this.props.fetchCategoryRecipes(id);
    }

    renderRecipes() {
        const { recipes } = this.props;
        if (!recipes) {
            return (
                <div>Loading...</div>
            );
        }

        if(_.size(recipes) === 0) {
            return (
                <div className="alert alert-info">
                    Category has no recipes in it yet. Click <strong>Add Recipe</strong> to add recipes
                </div>
            );
        }

        //using lodash map to traverse through the category object
        return _.map(recipes, recipe => {
            const imgSrc = recipeListImg;
            return (
                <div className="col-md-6" key={ recipe.id }>
                    <div className="card">
                        <img className="card-img-top" src={ imgSrc } />
                        <div className="card-body">
                            <h5 className="card-title">{ recipe.name }</h5>
                            <p className="card-text">
                                { recipe.description }
                            </p>
                            <Link to={`/categories/${recipe.cat_id}/recipes/${recipe.id}`} className="btn btn-primary">View More</Link>
                        </div>
                    </div>
                </div>
            );
        });
    }

    getPaginatedItems = (e) => {
        e.preventDefault();
        const { id } = this.props.match.params;
        this.props.fetchCategoryRecipes(id, e.target.dataset.page);
    }

    renderPagination = () => {
        const { pagination } = this.props;
        if(!pagination){
            return (
                <div>Loading pagination...</div>
            );
        }
        const prevClass = `${pagination.previous_page ? 'page-item' : 'page-item disabled'}`;
        const nextClass = `${pagination.next_page ? 'page-item' : 'page-item disabled'}`;
        return (
            <ul class="pagination justify-content-center">
                <li className={ prevClass }>
                    <a 
                        className="page-link" href="#" tabindex="-1" 
                        onClick = { this.getPaginatedItems } 
                        data-page={ pagination.previous_page }>Previous
                    </a>
                </li>
                <li className='page-item'>
                    <a className="page-link" href="#">
                        Page { pagination.current_page } of { pagination.pages }
                    </a>
                </li>
                <li className={ nextClass }>
                    <a 
                        className="page-link" href="#" 
                        onClick = { this.getPaginatedItems } 
                        data-page={ pagination.next_page }>Next
                    </a>
                </li>
            </ul>
        );
    }

    render() {
        return (
            <div>
                <div className="row">
                    { this.renderRecipes() }
                </div><br />

                {/* pagination */}
                <nav aria-label="...">
                    { this.renderPagination() }
                </nav>

                <div>
                    <RecipeModal reloadItems = { this.reloadRecipes }/>
                </div>
                <Notifications />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return { 
        recipes: state.category.category_recipes,
        pagination: state.category.recipePagination
    }
}

export default withRouter(connect(mapStateToProps, { fetchCategoryRecipes })(CategoryRecipes));