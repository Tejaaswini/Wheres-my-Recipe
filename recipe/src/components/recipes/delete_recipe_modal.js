import React, { Component } from 'react';
import { connect } from 'react-redux';
import Notifications from 'react-notify-toast';

import { deleteRecipe } from '../../actions/recipes'

class DeleteRecipeModal extends Component {

    // method to handle delete of recipe
    handleDelete = () => {
        const { recipe, redirect } = this.props;
        this.props.deleteRecipe(recipe.cat_id, recipe.id, () => {
            document.querySelector('#close').click();
            redirect();
        });
    }

    render () {
        return (
            <div>
                <div className="modal fade" id="deleteRecipe" tabindex="-1" role="dialog" aria-labelledby="deleteRecipeTitle" aria-hidden="true">
                    <div className="modal-dialog modal-dialog-centered" role="document">
                        <div className="modal-content">
                        <div className="modal-header">
                            <h5 className="modal-title" id="recipeModalLongTitle">Delet Recipe</h5>
                            <button type="button" className="close" data-dismiss="modal" aria-label="Close" id="close">
                            <span aria-hidden="true">&times;</span>
                            </button>
                        </div>
                        <div className="modal-body">
                            <div className="alert alert-danger">Are you sure you want to delete this recipe?</div>
                        </div>
                        <div className="modal-footer">
                            <button type="button" className="btn btn-secondary" data-dismiss="modal">Cancel</button>
                            <button onClick={ this.handleDelete } type="submit" className="btn btn-danger" form="recipeForm">Delete</button>
                        </div>
                        </div>
                    </div>
                </div>
                <Notifications />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return { recipe: state.recipe.recipe_item };
}

export default connect(mapStateToProps, { deleteRecipe })(DeleteRecipeModal);