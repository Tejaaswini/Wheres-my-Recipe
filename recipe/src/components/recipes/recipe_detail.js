import _ from 'lodash';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import Notifications from 'react-notify-toast';

import { fetchRecipe } from '../../actions/recipes';
import RecipeEditModal from './recipe_edit_modal';
import DeleteRecipeModal from './delete_recipe_modal';
import recipeListImg from '../../img/recipe_detail.png'

export class RecipeDetail extends Component {
    componentDidMount() {
        this.loadItem();
    }

    loadItem = () => {
        const { cat_id, recipe_id } = this.props.match.params;
        this.props.fetchRecipe(cat_id, recipe_id);
    }

    redirectTo = () => {
        const { cat_id } = this.props.match.params;
        this.props.history.push(`/categories/${cat_id}`)
    }

    renderItems(items) {
        return _.map(_.split(items, '\n'), item => {
            return (
                <li>
                    <i className="fa fa-plus-circle"></i> { item }
                </li>
            );
        });
    }

    render() {
        const { recipe } = this.props;
        const imgSrc = recipeListImg;
        if (!recipe){
            return (
                <div>Loading...</div>
            );
        }
        return (
            <div className="container">
                <div className="row">
                    <div className="col-md-12">
                        <div className="card">
                            <img className="card-img-top" src={ imgSrc } />
                            <div className="card-body">
                                <h5 className="card-title">
                                    { recipe.name } in category {'  '}   
                                    <small><a href={`/categories/${recipe.cat_id}`}>{ recipe.cat_name }</a></small>
                                </h5>
                                <p className="card-text">
                                    { recipe.description }
                                </p>
                            </div>
                        </div><br />

                        <button className="btn btn-danger pull-right" title="Delete Recipe" data-toggle="modal" data-target="#deleteRecipe">
                            <i className="fa fa-trash-o fa-fw"></i> 
                        </button>                        
                        <button className="btn btn-info" data-toggle="modal" data-target="#recipeModal">
                            <i className="fa fa-edit fa-fw"></i> Edit Recipe</button>
                        <br />
                        <hr />
                    </div>
                </div>

                <div className="row">
                    <div className="col-6">
                        <h5>Ingredients</h5>
                        <hr />
                        <ul class="list-unstyled">
                            { this.renderItems(recipe.ingredients) }
                        </ul>
                    </div>

                    <div className="col-6">
                        <h5>Directions/Steps</h5>
                        <hr />
                        <ul class="list-unstyled">
                            { this.renderItems(recipe.directions) }
                        </ul>
                    </div>
                </div>
                <div>
                    <RecipeEditModal reload = { this.loadItem }/>
                    <DeleteRecipeModal redirect = { this.redirectTo } />
                    <Notifications />
                </div>

                <hr className="featurette-divider" />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return { 
        recipe: state.recipe.recipe_item
    };
}

export default connect(mapStateToProps, { fetchRecipe })(RecipeDetail);
