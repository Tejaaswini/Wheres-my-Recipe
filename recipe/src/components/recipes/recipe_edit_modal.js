import React, { Component } from 'react';
import { connect } from 'react-redux';
import Notifications from 'react-notify-toast';

import { editRecipe } from '../../actions/recipes';

class RecipeEditModal extends Component {
    constructor(props) {
        super(props);
        this.state = { value: props.recipe.name }
    }

    handleInputChange = (event) => {
        this.setState({ value: event.target.value });
    }
    // submiting form values
    _submit = (values) => {
        const { recipe, reload } = this.props;
        this.props.editRecipe(values, recipe.cat_id, recipe.id, () => {
            document.querySelector('#close').click();
            reload();
        });
    }
    // handle form submition
    _onhandleSubmit = (event) => {
        event.preventDefault();
        const formData = {};
        for (const field in this.refs) {
            formData[field] = this.refs[field].value;
        }
        this._submit(formData);
    }

    renderModalForm() {
        const { recipe } = this.props;
        return (
            <div>
                <form id="recipeForm" onSubmit={ this._onhandleSubmit }>
                <div className="form-group row">
                    <label for="name" className="col-sm-3 col-form-label" name="name">Name</label>
                    <div className="col-sm-9">
                        <input ref="name" type="text" className="form-control" value={ this.state.value } id="name" onChange = { this.handleInputChange } />
                    </div>
                </div>
                <div class="form-group row">
                    <label for="description" className="col-sm-3 col-form-label">Description</label>
                    <div className="col-sm-9">
                        <textarea ref="description" className="form-control" id="description" name="description">{ recipe.description }</textarea>
                    </div>
                </div>
                <div class="form-group row">
                    <label for="ingredients" className="col-sm-3 col-form-label">Ingredients</label>
                    <div className="col-sm-9">
                        <textarea ref="ingredients" className="form-control" id="ingredients" name="ingredients">{ recipe.ingredients }</textarea>
                    </div>
                </div>
                <div class="form-group row">
                    <label for="directions" className="col-sm-3 col-form-label">Directions</label>
                    <div className="col-sm-9">
                        <textarea ref="directions" className="form-control" id="directions" name="directions">{ recipe.directions }</textarea>
                    </div>
                </div>
            </form>
            </div>
        );
    }

    render() {
        return (
            <div>
                <div className="modal fade" id="recipeModal" tabindex="-1" role="dialog" aria-labelledby="recipeModalTitle" aria-hidden="true">
                    <div className="modal-dialog modal-dialog-centered" role="document">
                        <div className="modal-content">
                        <div className="modal-header">
                            <h5 className="modal-title" id="recipeModalLongTitle">Edit Recipe</h5>
                            <button type="button" className="close" data-dismiss="modal" aria-label="Close" id="close">
                            <span aria-hidden="true">&times;</span>
                            </button>
                        </div>
                        <div className="modal-body">
                            { this.renderModalForm() }
                        </div>
                        <div className="modal-footer">
                            <button type="button" className="btn btn-secondary" data-dismiss="modal">Cancel</button>
                            <button type="submit" className="btn btn-info" form="recipeForm">Edit</button>
                        </div>
                        </div>
                    </div>
                </div>
                <Notifications />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return { recipe: state.recipe.recipe_item };
}

export default connect(mapStateToProps, { editRecipe })(RecipeEditModal);
