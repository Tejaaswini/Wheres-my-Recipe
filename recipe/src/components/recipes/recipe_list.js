import _ from 'lodash';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Notifications from 'react-notify-toast';
import recipeListImg from '../../img/recipe_list_image.jpg'

import { fetchRecipes } from '../../actions/recipes';

export class RecipesList extends Component {

    componentDidMount() {
        this.props.fetchRecipes();
    }

    onInputChange = (event) => {
        const { value } = event.target
        this.props.fetchRecipes(value);
    }

    getPaginatedItems = (e) => {
        e.preventDefault();
        const value='';
        this.props.fetchRecipes(value, e.target.dataset.page);
    }

    renderPagination = () => {
        const { pagination } = this.props;
        if(!pagination){
            return (
                <div>Loading pagination...</div>
            );
        }

        if(pagination.total_count === 0) {
            return (
                <div className="alert alert-warning">
                    No recipes found that match this search term!
                </div>
            );
        }

        const prevClass = `${pagination.previous_page ? 'page-item' : 'page-item disabled'}`;
        const nextClass = `${pagination.next_page ? 'page-item' : 'page-item disabled'}`;
        return (
            <ul class="pagination justify-content-center">
                <li className={ prevClass }>
                    <a 
                        className="page-link" href="#" tabindex="-1" 
                        onClick = { this.getPaginatedItems } 
                        data-page={ pagination.previous_page }>Previous
                    </a>
                </li>
                <li className='page-item'>
                    <a className="page-link" href="#">
                        Page { pagination.current_page } of { pagination.pages }
                    </a>
                </li>
                <li className={ nextClass }>
                    <a 
                        className="page-link" href="#" 
                        onClick = { this.getPaginatedItems } 
                        data-page={ pagination.next_page }>Next
                    </a>
                </li>
            </ul>
        );
    }

    renderRecipes() {
        const { recipes } = this.props;
        if (!recipes) {
            return (
                <div>Loading...</div>
            );
        }

        if(_.size(recipes) === 0) {
            return (
                <div className="alert alert-info">
                    You do not have recipes yet, first add categories and then recipes
                </div>
            );
        }

        //using lodash map to traverse through the category object
        return _.map(recipes, recipe => {
            const imgSrc = recipeListImg;
            return (
                <div className="col-md-6" key={ recipe.id }>
                    <div className="card">
                        <img className="card-img-top" src={ imgSrc } />
                        <div className="card-body">
                            <h5 className="card-title">{ recipe.name }</h5>
                            <p className="card-text">{ recipe.description }</p>
                            <Link to={`/categories/${recipe.cat_id}/recipes/${recipe.id}`} className="btn btn-primary">View More</Link>
                        </div>
                    </div>
                </div>
            );
        });
    }

    render() {
        return (
            <div>
                <div className="input-group mb-3 col-md-6">
                    <input 
                        type="text" 
                        className="form-control"  
                        onChange = { this.onInputChange }
                        placeholder="Search Recipe" 
                        aria-label="Recipe" 
                        aria-describedby="basic-addon2" 
                    />
                    {/* <div className="input-group-append">
                        <span className="input-group-text" id="basic-addon2"><i className="fa fa-search"></i> Search</span>
                    </div> */}
                </div>
                <div className="row">
                    { this.renderRecipes() }
                </div><br />

                {/* pagination */}
                <nav aria-label="...">
                    { this.renderPagination() }
                </nav>
                <Notifications />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return { 
        recipes: state.recipes.recipes,
        pagination: state.recipes.recipePagination 
    }
}

export default connect(mapStateToProps, { fetchRecipes })(RecipesList);