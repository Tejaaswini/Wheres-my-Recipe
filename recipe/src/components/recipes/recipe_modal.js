import React, { Component } from 'react';
import { Field, reduxForm, reset } from 'redux-form';
import {Col, Form, FormGroup, FormFeedback, FormText, Label, Input } from 'reactstrap';
import { connect } from 'react-redux';
import Notifications from 'react-notify-toast';

import { createRecipe } from '../../actions/recipes';

class RecipeModal extends Component {

    renderField(field) {
        const { meta: { touched, error } } = field;
        const isValid = touched && error ? false : '';
        return (
            <FormGroup row>
                <Label htmlFor={ field.htmlfor } md={3}>{ field.label }</Label>
                <Col md={9}>
                    <FormText color="muted">
                        { field.text }
                    </FormText>
                    <Input 
                        ref = { field.name }
                        type={ field.type }
                        id={ field.id }
                        { ...field.input }
                        valid = { isValid }
                    />
                    <FormFeedback>{ touched ? error : '' }</FormFeedback>
                </Col>
            </FormGroup>
        );
    }

    onSubmitForm = (values) => {
        const { category, reloadItems } = this.props;
        this.props.createRecipe(values, category.id, () => {
            document.querySelector('#close').click();
            reloadItems();
        });
    }

    renderModalForm() {
        const { handleSubmit } = this.props;
        return (
            <div>
                <Form onSubmit={ handleSubmit(this.onSubmitForm.bind(this)) } id="recipeForm" name="recipeForm">
                    <Field 
                        label="Recipe name"
                        name="name"
                        htmlfor="recipe_name"
                        id="recipe_name"
                        type="text"
                        component={ this.renderField }
                    />
                    <Field 
                        label="Description"
                        name="description"
                        htmlfor="description"
                        id="description"
                        type="textarea"
                        component={ this.renderField }
                    />
                    <Field 
                        label="Ingredients"
                        name="ingredients"
                        htmlfor="ingredients"
                        id="iingredients"
                        type="textarea"
                        text="Type with space and each on new line."
                        component={ this.renderField }
                    />
                    <Field 
                        label="Directions"
                        name="directions"
                        htmlfor="directions"
                        id="directions"
                        type="textarea"
                        text="Type with space and each on new line."
                        component={ this.renderField }
                    />
                </Form>
            </div>
        );
    }

    render() {
        return (
            <div>
                <div className="modal fade" id="recipeModal" tabindex="-1" role="dialog" aria-labelledby="recipeModalTitle" aria-hidden="true">
                    <div className="modal-dialog modal-dialog-centered" role="document">
                        <div className="modal-content">
                        <div className="modal-header">
                            <h5 className="modal-title" id="recipeModalLongTitle">Create New Recipe</h5>
                            <button type="button" className="close" data-dismiss="modal" aria-label="Close" id="close">
                            <span aria-hidden="true">&times;</span>
                            </button>
                        </div>
                        <div className="modal-body">
                            { this.renderModalForm() }
                        </div>
                        <div className="modal-footer">
                            <button type="button" className="btn btn-secondary" data-dismiss="modal">Cancel</button>
                            <button type="submit" className="btn btn-success" form="recipeForm">Save</button>
                        </div>
                        </div>
                    </div>
                </div>
                <Notifications />
            </div>
        );
    }
}

// validate input form field from values
const validate = (values) => {
    const errors = {};
    if (!values.recipe_name){
        errors.recipe_name = 'Name is required!';
    }

    if (!values.description) {
        errors.description = 'Type Description!'
    }

    if (!values.ingredients) {
        errors.ingredients = 'Recipe must have ingredients'
    }

    if (!values.directions) {
        errors.directions = 'Directions must be filled!'
    }

    return errors
}

const afterSubmit = (result, dispatch) =>
  dispatch(reset('RecipeForm'));


const mapStateToProps = (state) => {
    return { category: state.category.category_item };
}

const RecipeForm =  reduxForm({
    validate,
    form: 'RecipeForm',
    onSubmitSuccess: afterSubmit
})(
    connect(mapStateToProps, { createRecipe })(RecipeModal)
);

export default RecipeForm;