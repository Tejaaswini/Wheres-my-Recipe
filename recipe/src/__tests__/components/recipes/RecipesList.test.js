import _ from 'lodash';
import React from 'react';
import ReactDOM from 'react-dom';
import configureStore from 'redux-mock-store';
import reduxThunk from 'redux-thunk';
import renderer from 'react-test-renderer';
import { connect } from 'react-redux';
import { Provider } from 'react-redux';
import { Link } from 'react-router-dom';
import {configure, mount, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16'
import { BrowserRouter as Router } from 'react-router-dom';
import { RecipesList } from '../../../components/recipes/recipe_list';
import mockData from '../../../__mocks__/mockData';

const middlewares = [reduxThunk];
const mockStore = configureStore(middlewares);

configure({adapter: new Adapter()});

describe('RecipesList', () => {
    const { recipes } = mockData;
    let store, wrapper;
    let state = {
        auth: {
            authenticated: true
        },
        categories: {},
        category: {},
        recipes: recipes,
        recipe: {},
    };

    const props = {
        categories: state.recipes,
        fetchRecipes: () => Promise.resolve()
    }

    beforeEach(() => {
        store = mockStore(state);
    });
    it('it renders without crashing', () => {
        mount(
            <Provider store={store}>
                <Router>
                    <RecipesList { ...props } />
                </Router>
            </Provider>
        );
    });

    it('it has valid snapshot', () => {
        const component = renderer.create(
            <Provider store={store}>
                <Router>
                    <RecipesList { ...props } />
                </Router>
            </Provider>
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});