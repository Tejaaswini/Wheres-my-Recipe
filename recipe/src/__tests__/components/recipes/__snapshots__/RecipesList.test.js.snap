// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipesList it has valid snapshot 1`] = `
<div>
  <div
    className="input-group mb-3 col-md-6"
  >
    <input
      aria-describedby="basic-addon2"
      aria-label="Recipe"
      className="form-control"
      onChange={[Function]}
      placeholder="Search Recipe"
      type="text"
    />
    <div
      className="input-group-append"
    >
      <span
        className="input-group-text"
        id="basic-addon2"
      >
        <i
          className="fa fa-search"
        />
         Search
      </span>
    </div>
  </div>
  <div
    className="row"
  >
    <div>
      Loading...
    </div>
  </div>
  <br />
  <nav
    aria-label="..."
  >
    <div>
      Loading pagination...
    </div>
  </nav>
  <div
    id="notification-wrapper"
  />
</div>
`;
